### PR TITLE
fix: rng handling with workers

### PR DESCRIPTION
**Pull request by selmanozleyen**
hi @ilan-gold @flying-sheep ,

The current code doesn't use spawn and uses kind of unconventional worker rng handling. For example this line ensures each worker returns the same rng for each worker because it was used to sharing chunks across workers https://github.com/scverse/annbatch/blob/9bbaa8b43ef48b2805fb59fe3369d511f6f52c11/src/annbatch/utils.py#L127-L129 

Another problem is https://github.com/scverse/annbatch/blob/9bbaa8b43ef48b2805fb59fe3369d511f6f52c11/src/annbatch/samplers/_chunk_sampler.py#L164-L166 where you can't control reproducilibity for batch shuffling. I added a unit test here which you can cherry-pick https://github.com/scverse/annbatch/tree/1bcbb8887e5042c52e0c21dc3ad327c1a1a15bc1 and see it fails on main but passes on this fix (the other changes I made to the file are to update the other tests with MockWorkerHandle's)

Here is the proposed way to handle rngs now

- For batch shuffling I use the worker_rng because usually each worker gets a different chunks and should shuffle their batches independently
- For chunk sharing/shuffling I use the ChunkSampler._rng because ChunkSampler is created before the torch workers are spawned and the copy of ChunkSampler's will use the same ChunkSampler._rng and guarantee that the chunks are shared the same way without overlapping.

why we should use spawn: https://numpy.org/doc/stable/reference/random/multithreading.html

Note: we can talk about the details of `WorkerHandler.__init__`, it depends on the expected usage of it I guess but I just wanted to show what I was talking about in the standup

### CODE REVIEW

File: tests/test_sampler.py [hunk 1]
```diff
@@ -26,28 +26,38 @@ def collect_indices(sampler, n_obs):
...
+    def __init__(self, worker_id: int, num_workers: int, rng: np.random.Generator | None = None):
         self.worker_id = worker_id
         self._num_workers = num_workers
-        self._rng = np.random.default_rng(seed)
+        # Each worker gets its own RNG spawned from the sampler's RNG (mirrors real WorkerHandle)
+        if rng is not None:
+            bit_generators = rng.bit_generator.spawn(num_workers)
+        else:
+            bit_generators = np.random.SeedSequence(42).spawn(num_workers)
+        self._rng = np.random.default_rng(bit_generators[worker_id])
```
Comment (flying-sheep) [inline tests/test_sampler.py]: I don’t get this. Why do you spawn `num_workers` RNGs and then take one and throw the others away?

If that behavior makes sense, it should look like this, i.e.
- no need to go through `.bit_generator`, `Generator.spawn()` works fine.
- `rng=None` conventionally means no seed / total randomness:

```suggestion
        if rng is not None:
            generators = rng.spawn(num_workers)
        else:  # passing no `rng` means you want total randomness, don’t set a seed here.
            generators = np.random.default_rng().spawn(num_workers)
        self._rng = generators[worker_id]
```

File: src/annbatch/utils.py [hunk 2]
```diff
@@ -102,44 +102,50 @@ def __iter__(self):
...
+        if rng is not None:
+            bit_generators = rng.bit_generator.spawn(self._worker_info.num_workers)
         else:
-            # This is used for the _get_chunks function
-            # Use the same seed for all workers that the resulting splits are the same across workers
-            # torch default seed is `base_seed + worker_id`. Hence, subtract worker_id to get the base seed
-            return np.random.default_rng(self._worker_info.seed - self._worker_info.id)
+            seq = np.random.SeedSequence(self._worker_info.seed).spawn(self._worker_info.num_workers)
+            bit_generators = seq
+        self._rng = np.random.default_rng(bit_generators[self._worker_info.id])
```
Comment (flying-sheep) [inline src/annbatch/utils.py]: ditto. also deduplicate.

**Comment by selmanozleyen** [comment]
> OK, I think you’re making things too complicated. Why not just delete WorkerHandle and do

I agree actually. I wanted to work with the old choices but I guess `WorkerHandle` doesn't make sense with Sampler API anyway.

File: src/annbatch/samplers/_chunk_sampler.py [hunk 3]
```diff
@@ -119,50 +139,37 @@ def validate(self, n_obs: int) -> None:
...
-        if (
-            worker_handle is not None
-            and worker_handle.num_workers > 1
-            and not self._drop_last
-            and self._batch_size != 1
-        ):
+
+        if worker_info is not None and worker_info.num_workers > 1 and not self._drop_last and self._batch_size != 1:
+            # With batch_size=1, every batch is exactly 1 item, so no partial batches exist
             raise ValueError("When using DataLoader with multiple workers drop_last=False is not supported.")
```
[the remark this answers is not in the document]
Comment (selmanozleyen) [reply src/annbatch/samplers/_chunk_sampler.py]: > Just out of curiosity, could we in theory drop_last per-worker?

Not with `ChunkSampler` because we guarantee only last chunk can be incomplete so only one worker drops it. But in my `CategoricalSampler` PR for example, this could've happen if I implement it with multiple workers. If each worker had one category then each would have to drop because each workers last chunk could've been incomplete.

> Is there a way to discover if our data loader is being run in a context with custom collation

I don't think it currently can run with the default collation_fn. I think for torch it's just a function even if it's default one.

> Judging by [torch's docs](https://docs.pytorch.org/docs/stable/data.html#multi-process-data-loading):

I don't think this applies to us because if we where to give our Loader we probably better use batch_size=None in torch.DataLoader. It only applies to whatever batch_size we give to the torch.DataLoader not the batch_sizes we yield.

File: src/annbatch/utils.py [hunk 4]
```diff
@@ -102,59 +101,20 @@ def __iter__(self):
                 total += gap
 
 
-class WorkerHandle:  # noqa: D101
-    @cached_property
-    def _worker_info(self):
-        if find_spec("torch"):
-            from torch.utils.data import get_worker_info
+def _spawn_worker_rng(
+    rng: np.random.Generator | None,
```
[the remark this answers is not in the document]
Comment (ilan-gold) [reply src/annbatch/utils.py]: But the lines are uncovered at the moment: https://app.codecov.io/gh/scverse/annbatch/pull/135?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=scverse.  That seems bad, like we're writing code purely for ourselves that we also don't use?

File: tests/test_sampler.py [hunk 5]
```diff
@@ -195,6 +170,40 @@ def test_workers_cover_full_dataset_without_overlap(
     assert set().union(*all_worker_indices) == set(range(n_obs))
 
 
+def test_batch_shuffle_is_reproducible():
+    """Test that batch shuffling is reproducible when using ChunkSampler directly."""
```
Comment (ilan-gold) [inline tests/test_sampler.py]:
```suggestion
def test_batch_shuffle_is_reproducible_with_same_seed_rng():
    """Test that batch shuffling is reproducible when passing in rngs with identical seeds to ChunkSampler directly."""
```

**Review by ilan-gold — COMMENTED** [review]
Can you add a test for this that would fail on main? If we're fixing something, I would guess something is broken?

File: src/annbatch/samplers/_chunk_sampler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ilan-gold) [reply src/annbatch/samplers/_chunk_sampler.py]: Ok I trust you! What you say makes sense but this interaction is overall brittle (hence my desire not to dump time into this without a concrete use-case).  For example:

> I don't think this applies to us because if we where to give our Loader we probably better use batch_size=None in torch.DataLoader

We don't do anything, our users do.  What if they use `batch_size!=None`?

File: src/annbatch/samplers/_chunk_sampler.py [hunk 6]
```diff
@@ -119,50 +139,34 @@ def validate(self, n_obs: int) -> None:
...
-        if worker_handle is not None:
-            chunks = worker_handle.get_part_for_worker(chunks)
+        if worker_info is not None:
+            chunks = np.array_split(chunks, worker_info.num_workers)[worker_info.id]
         # Set up the iterator for chunks and the batch indices for splits
         in_memory_size = self._chunk_size * self._preload_nchunks
         chunks_per_request = split_given_size(chunks, self._preload_nchunks)
         batch_indices = np.arange(in_memory_size)
         split_batch_indices = split_given_size(batch_indices, self._batch_size)
+        batch_rng = _spawn_worker_rng(self._rng, worker_info.id) if worker_info else self._rng
```
Comment (flying-sheep) [inline src/annbatch/samplers/_chunk_sampler.py]: again, why define a `_spawn_worker_rng` instead of just using `np.random.default_rng(worker_info.seed)`?

**Comment by selmanozleyen** [comment]
@ilan-gold @flying-sheep even if we were to use torch.seed we would need to use it in the root ChunkSampler._rng level again since we want everything to be reproducable by torch.seed. 

If we want explicit torch support, if no rng is provided we can create the base rng with torch.seed if there is torch present. But I think it's better to explicitly document: "This Loader doesn't use the torch random number generator therefore setting torch.seed will not effect the reproducilibity. For reproducibility see ..."

**Review by ilan-gold — COMMENTED** [review]
Can you add this behavior explanation to the docs along with a warning that using `torch` is neither encouraged nor guaranteed to behave as people expect?

**Review by ilan-gold — CHANGES_REQUESTED** [review]

**Review by flying-sheep — APPROVED** [review]
Looks good!

**Review by ilan-gold — APPROVED** [review]
